lib/semantics/check-omp-structure.cc

```diff
@@ -114,18 +113,57 @@ void OmpStructureChecker::Enter(const parser::OmpLoopDirective::Do &) {
   // reserve for nesting check
   SetContextDirectiveEnum(OmpDirective::DO);
 
-  OmpClauseSet allowed{OmpClause::PRIVATE, OmpClause::LASTPRIVATE,
-      OmpClause::LINEAR, OmpClause::REDUCTION, OmpClause::SCHEDULE,
-      OmpClause::COLLAPSE, OmpClause::ORDERED};
+  OmpClauseSet allowed{OmpClause::PRIVATE, OmpClause::FIRSTPRIVATE,
+      OmpClause::LASTPRIVATE, OmpClause::LINEAR, OmpClause::REDUCTION};
   SetContextAllowed(allowed);
 
   OmpClauseSet allowedOnce{
       OmpClause::SCHEDULE, OmpClause::COLLAPSE, OmpClause::ORDERED};
   SetContextAllowedOnce(allowedOnce);
 }
 
+void OmpStructureChecker::Leave(const parser::OmpClauseList &) {
+  // 2.7 Loop Construct Restriction
+  if (GetContext().directive == OmpDirective::DO) {
+    if (SeenClause(OmpClause::SCHEDULE)) {
+      // only one schedule clause is allowed
+      auto it{GetContext().clauseInfo.find(OmpClause::SCHEDULE)};
+      auto *clause{it->second};
```

Comment:
You could replace `SeenClause()` with `FindClause()` that returns `const parser::OmpClause *`.
Then here you could check if the SCHEDULE clause is present and get it all in one step.

Reading this code, it is surprising that you don't check that `it` is valid (i.e. not `end()`) before dereferencing it.

Reply:
Oops... I will fix that, thanks!

Reply:
Addressed in commit [1b57193](https://github.com/flang-compiler/f18/pull/536/commits/1b57193d2377c5121f16075950b64ea77fa72e2e).